osu.Game/Screens/Play/SaveFailedScoreButton.cs

```diff
@@ -0,0 +1,140 @@
+        {
+            InternalChild = shakeContainer = new ShakeContainer
+            {
+                RelativeSizeAxes = Axes.Both,
+                Child = button = new DownloadButton
+                {
+                    RelativeSizeAxes = Axes.Both,
+                }
+            };
+
+            button.Action = () =>
+            {
+                switch (State.Value)
+                {
+                    case ImportState.Imported:
+                        game?.PresentScore(score, ScorePresentType.Gameplay);
+                        break;
+
+                    case ImportState.Importing:
+                        shakeContainer.Shake();
+                        break;
+
+                    default:
+                        saveScore();
+                        break;
+                }
+            };
+            State.BindValueChanged(state =>
+            {
+                switch (state.NewValue)
+                {
+                    case ImportState.Imported:
+                        button.State.Value = DownloadState.LocallyAvailable;
+                        break;
+
+                    case ImportState.Importing:
+                        button.State.Value = DownloadState.Importing;
+                        break;
+
+                    case ImportState.Failed:
+                        button.State.Value = DownloadState.NotDownloaded;
+                        break;
+                }
+            }, true);
+            State.BindValueChanged(updateTooltip, true);
+        }
+
+        private void saveScore()
+        {
+            State.Value = ImportState.Importing;
+            saveReplayAsync = Task.Run(SaveReplay);
+
+            saveScoreDelegate = new ScheduledDelegate(() =>
+            {
+                if (saveReplayAsync?.IsCompleted != true)
+                    // If the asynchronous preparation has not completed, keep repeating this delegate.
+                    return;
+
+                saveScoreDelegate?.Cancel();
+
+                score = saveReplayAsync.GetAwaiter().GetResult();
+
+                State.Value = score != null ? ImportState.Imported : ImportState.Failed;
+            }, Time.Current, 50);
+
+            Scheduler.Add(saveScoreDelegate);
+        }
+
+        private void updateTooltip(ValueChangedEvent<ImportState> state)
+        {
+            switch (state.NewValue)
+            {
+                case ImportState.Imported:
+                    button.TooltipText = @"Watch replay";
+                    break;
+
+                case ImportState.Importing:
+                    button.TooltipText = @"Importing score";
+                    break;
+
+                case ImportState.Failed:
+                    button.TooltipText = @"Import failed, click button to re-import";
+                    break;
+
+                default:
+                    button.TooltipText = @"Save score";
+                    break;
+            }
+        }
+
+        public enum ImportState
```

Comment:
I don't think this should be required.

Reply:
What I originally intended to use was `DownloadState`
but DownloadState doesn't have `ImportFailed` State so I create a new enum here

in fact `DownloadState` need a `failed` stat?
The original button will revert to `NotDownloaded` when the download fails

Reply:
What's the case where it can fail?

Reply:
May only have a hard disk write issue in the current version and may be able to be resolved by re-importing
in past versions (before https://github.com/ppy/osu/pull/19025), custom rulesets also fail to import, and repeating the import doesn't help